File: src/main/java/org/khun/leetcode/ArrayAndString/RotateArray.java
package org.khun.leetcode.ArrayAndString;

import java.util.Arrays;

public class RotateArray {
    public void rotate(int[] nums, int k) {
        reverse(nums, 0, nums.length - 1);
        reverse(nums, 0, k - 1);
        reverse(nums, k, nums.length - 1);
    }

    public void reverse(int[] nums, int start, int end) {
        while (start < end) {
            int temp = nums[start];
            nums[start] = nums[end];
            nums[end] = temp;
            start++;
            end--;
        }
    }

    public static void main(String[] args) {
        RotateArray r = new RotateArray();

        int[] arr = {1, 2, 3, 4, 5, 6, 7};
        r.rotate(arr, 3);
        System.out.println(Arrays.toString(arr));

        int[] arr2 = {-1, -100, 3, 99};
        r.rotate(arr2, 2);
        System.out.println(Arrays.toString(arr2));
    }
}

File: src/main/java/org/khun/leetcode/ArrayAndString/RotateArray.kt
package org.khun.leetcode.ArrayAndString

fun rotateArray(nums: IntArray, k: Int) {
    reverseArray(nums, 0, nums.size - 1)
    reverseArray(nums, 0, k - 1)
    reverseArray(nums, k, nums.size - 1)
}

fun reverseArray(nums: IntArray, startIndex: Int, endIndex: Int) {
    var start = startIndex
    var end = endIndex
    while (start < end) {
        val temp = nums[start]
        nums[start] = nums[end]
        nums[end] = temp
        start++
        end--
    }
}

fun main() {
    val arr = intArrayOf(1, 2, 3, 4, 5, 6, 7)
    rotateArray(arr, 3)
    println(arr.contentToString())

    val arr2 = intArrayOf(-1, -100, 3, 99)
    rotateArray(arr2, 2)
    println(arr2.contentToString())
}